File: practical3.cpp
#include <iostream>
#include <fstream>
#include <unordered_set>
#include <vector>
#include <algorithm>
#include <cctype>
using namespace std;

unordered_set<string> keywords = {
    "int","long","void","char","return",
    "if","else","while","for"
};

unordered_set<char> operators = {
    '+','-','*','/','=','<','>','&','%'
};

unordered_set<char> punctuations = {
    '(',')','{','}',';',','
};

bool isNumber(const string &s) {
    for(char c:s)
        if(!isdigit(c)) return false;
    return true;
}

int main() {
    ifstream file("Practical3_input.c");
    if(!file) {
        cout << "Error opening file\n";
        return 0;
    }

    vector<string> symbolTable;
    vector<string> lexicalErrors;

    string line;
    int lineNo = 0;
    bool inBlockComment = false;

    cout << "TOKENS\n";

    while(getline(file, line)) {
        lineNo++;

        for (int i = 0; i < line.length(); i++) {

            
            if (inBlockComment) {
                if (i + 1 < line.length() && line[i] == '*' && line[i + 1] == '/') {
                    inBlockComment = false;
                    i++;
                }
                continue;
            }

            
            if (i + 1 < line.length() && line[i] == '/' && line[i + 1] == '*') {
                inBlockComment = true;
                i++;
                continue;
            }

            
            if (i + 1 < line.length() && line[i] == '/' && line[i + 1] == '/') {
                break;
            }

            if (isspace(line[i])) continue;

            
            if (isalpha(line[i]) || line[i] == '_') {
                string word;
                while (i < line.length() && (isalnum(line[i]) || line[i] == '_'))
                    word += line[i++];

                i--;

                if (keywords.count(word))
                    cout << "Keyword: " << word << endl;
                else {
                    cout << "Identifier: " << word << endl;
                    if (find(symbolTable.begin(), symbolTable.end(), word) == symbolTable.end())
                        symbolTable.push_back(word);
                }
            }

            
            else if (isdigit(line[i])) {
                string num;
                while (i < line.length() && (isalnum(line[i]) || line[i] == '.'))
                    num += line[i++];

                i--;

                if (isNumber(num))
                    cout << "Constant: " << num << endl;
                else
                    lexicalErrors.push_back(
                        "Line " + to_string(lineNo) + " : " + num + " invalid lexeme"
                    );
            }

            
            else if (line[i] == '"') {
                string str = "\"";
                i++;
                while (i < line.length() && line[i] != '"')
                    str += line[i++];
                str += "\"";
                cout << "Constant: " << str << endl;
            }

            
            else if (operators.count(line[i])) {
                cout << "Operator: " << line[i] << endl;
            }

            
            else if (punctuations.count(line[i])) {
                cout << "Punctuation: " << line[i] << endl;
            }
        }
    }

    cout << "\nLEXICAL ERRORS\n";
    for (auto &e : lexicalErrors)
        cout << e << endl;

    cout << "\nSYMBOL TABLE ENTRIES\n";
    for (int i = 0; i < symbolTable.size(); i++)
        cout << i + 1 << ") " << symbolTable[i] << endl;

    return 0;
}

File: practical3.py
import string

# Keywords, operators, punctuations
keywords = {
    "int", "long", "void", "char", "return",
    "if", "else", "while", "for"
}

operators = {'+', '-', '*', '/', '=', '<', '>', '&', '%'}
punctuations = {'(', ')', '{', '}', ';', ','}


def is_number(s):
    return s.isdigit()


def lexical_analyzer(filename):
    try:
        file = open(filename, "r")
    except:
        print("Error opening file")
        return

    symbol_table = []
    lexical_errors = []

    line_no = 0
    in_block_comment = False

    print("TOKENS")

    for line in file:
        line_no += 1
        i = 0

        while i < len(line):

            # Inside block comment
            if in_block_comment:
                if i + 1 < len(line) and line[i] == '*' and line[i + 1] == '/':
                    in_block_comment = False
                    i += 2
                else:
                    i += 1
                continue

            # Start block comment
            if i + 1 < len(line) and line[i] == '/' and line[i + 1] == '*':
                in_block_comment = True
                i += 2
                continue

            # Single-line comment
            if i + 1 < len(line) and line[i] == '/' and line[i + 1] == '/':
                break

            if line[i].isspace():
                i += 1
                continue

            # Identifier or keyword
            if line[i].isalpha() or line[i] == '_':
                word = ""
                while i < len(line) and (line[i].isalnum() or line[i] == '_'):
                    word += line[i]
                    i += 1

                if word in keywords:
                    print(f"Keyword: {word}")
                else:
                    print(f"Identifier: {word}")
                    if word not in symbol_table:
                        symbol_table.append(word)
                continue

            # Numeric constant
            if line[i].isdigit():
                num = ""
                while i < len(line) and (line[i].isalnum() or line[i] == '.'):
                    num += line[i]
                    i += 1

                if is_number(num):
                    print(f"Constant: {num}")
                else:
                    lexical_errors.append(
                        f"Line {line_no} : {num} invalid lexeme"
                    )
                continue

            # String constant
            if line[i] == '"':
                s = '"'
                i += 1
                while i < len(line) and line[i] != '"':
                    s += line[i]
                    i += 1
                s += '"'
                print(f"Constant: {s}")
                i += 1
                continue

            # Operator
            if line[i] in operators:
                print(f"Operator: {line[i]}")
                i += 1
                continue

            # Punctuation
            if line[i] in punctuations:
                print(f"Punctuation: {line[i]}")
                i += 1
                continue

            i += 1

    print("\nLEXICAL ERRORS")
    for err in lexical_errors:
        print(err)

    print("\nSYMBOL TABLE ENTRIES")
    for idx, sym in enumerate(symbol_table, start=1):
        print(f"{idx}) {sym}")

    file.close()


# Run analyzer
lexical_analyzer("testcase.c")
